File: python3/buddy_strings.py
"""
Space   : O(n)
Time    : O(n)
"""

from collections import Counter


class Solution:
    def buddyStrings(self, A: str, B: str) -> bool:
        dif = 0
        dupe = False
        ca, cb = Counter(A), Counter(B)
        n = len(A)

        # A must not be empty
        if n == 0:
            return False
        # A and B must have the same length
        if n != len(B):
            return False
        # A and B must have the same chars
        if len(ca - cb) > 0:
            return False

        # check if there's duplicate letter in A
        for k, v in ca.items():
            if v > 1:
                dupe = True
                break

        # get difference
        for idx in range(n):
            if A[idx] != B[idx]:
                dif += 1

        if dif == 0 and dupe:
            return True
        elif dif != 2:
            return False

        return True

File: golang/buddy_strings.go
package main

/*
Space	: O(n)
Time	: O(n)
*/

func buddyStrings(s string, goal string) bool {
	n := len(s)

	if n != len(goal) {
		return false
	}

	h1, h2 := map[string]int{}, map[string]int{}
	dif := 0
	for i := 0; i < n; i++ {
		if s[i] != goal[i] {
			dif += 1
		}

		if _, ok := h1[string(s[i])]; ok {
			h1[string(s[i])]++
		} else {
			h1[string(s[i])] = 1
		}

		if _, ok := h2[string(goal[i])]; ok {
			h2[string(goal[i])]++
		} else {
			h2[string(goal[i])] = 1
		}
	}

	multi := false
	for k, v := range h1 {
		if _, ok := h2[k]; ok {
			if h2[k] != v {
				return false
			}
		} else {
			return false
		}

		if v > 1 {
			multi = true
		}
	}

	if multi && dif <= 2 {
		return true
	}

	if dif == 2 {
		return true
	}

	return false
}
